[CLD-9186] Add framework for new Cloud Preview Modal

nickmisasi opened the pull request:
#### Summary
Adds a framework for new Cloud Preview intro modal. This modal consists of "slides" containing a SKU name, title, description, and video URL. We can easily add and remove "slides" by adding or removing JSON objects like below:

```
{
        skuLabel: 'ENTERPRISE ADVANCED',
        title: 'Welcome to your Mattermost preview',
        subtitle: 'This hands-on, 1-hour Mattermost Enterprise Advanced lets your team explore secure, mission-critical collaboration. The workspace is preloaded with <strong>sample data</strong> to get started.',
        videoUrl: 'https://www.youtube.com/watch?v=Zpyy2FqGotM', // TODO: Add proper video URL when available
        useCase: 'missionops'
}
```

I've added the 21 slides from the Figma designs. For now, you'll only be able to see the missionops ones.

Temoporarily, the modal is displayed on every load of the webapp for a Cloud Preview workspace. In a future PR when I incorporate the dynamic use case selection, this will be changed to stay dismissed upon dismissal.

#### Ticket Link
https://mattermost.atlassian.net/browse/CLD-9235

#### Screenshots
![image](https://github.com/user-attachments/assets/19f399c7-db6b-4611-a37c-f6910778c467)
![image](https://github.com/user-attachments/assets/1c5804e0-8d68-4102-b441-f10f6ac4b3cc)

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->
```release-note
None
```

nickmisasi commented:
/update-branch

mattermost-build commented:
Error trying to update the PR.
Please do it manually.

nickmisasi commented:
@claude I'm seeing webapp test failures. Here is an excerpt from the logs:

```
Summary of all failing tests
FAIL src/components/cloud_preview_modal/preview_modal_content.test.tsx
  ● PreviewModalContent › should not render SKU label when not provided

    expect(element).not.toBeInTheDocument()

    expected document not to contain element, found <span class="preview-modal-content__sku-label-logo" data-testid="mattermost-logo" /> instead

      55 |
      56 |         expect(screen.queryByText('ENTERPRISE')).not.toBeInTheDocument();
    > 57 |         expect(screen.queryByTestId('mattermost-logo')).not.toBeInTheDocument();
         |                                                             ^
      58 |     });
      59 |
      60 |     it('should render video when videoUrl is provided with .mp4 extension', () => {

      at Object.toBeInTheDocument (src/components/cloud_preview_modal/preview_modal_content.test.tsx:57:61)


Test Suites: 1 failed, 1 skipped, 905 passed, 906 of 907 total
Tests:       1 failed, 10 skipped, 7986 passed, 7997 total
Snapshots:   1441 passed, 1441 total
Time:        726.788 s
Ran all test suites.
npm ERR! Lifecycle script `test-ci` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: mattermost-webapp@10.6.0 
npm ERR!   at location: /home/runner/work/mattermost/mattermost/webapp/channels 

```

Please identify and fix the problem with the tests.

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/cloud_preview_modal.test.tsx:
Could you rewrite these tests to use React Testing Library instead of Enzyme? We shouldn't be adding any new Enzyme tests since we're going to need to remove Enzyme later this year

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.tsx:
```diff
@@ -0,0 +1,122 @@
+}
+
+// Helper function to extract YouTube video ID from various YouTube URL formats
+const getYouTubeVideoId = (url: string): string | null => {
```
Can this be replaced with `getVideoId` from `utils/youtube`?

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.tsx:
```diff
@@ -0,0 +1,122 @@
+            <h2 className='preview-modal-content__title'>{content.title}</h2>
+            <div
+                className='preview-modal-content__subtitle'
+                dangerouslySetInnerHTML={{__html: content.subtitle}}
```
Should this be using `dangerouslySetInnerHTML`? That seems like a mistake

hmhealey replied:
If we're doing that in the future, we should probably look at either iframes embedding content directly from the CWS or using Markdown. Perhaps it's just the prop name that's getting to me, but injecting HTML from somewhere else seems dangerous 😅

nickmisasi replied:
Yeah in the future if we do go for request-based data I'll definitely pivot to markdown.

esarafianou replied:
@nickmisasi what do you mean that you removed it in the next diff? I still see it in the code.

nickmisasi replied:
@esarafianou it was removed here: https://github.com/mattermost/mattermost/pull/31235/commits/454ba232f360f6ff9f252a3f7268da8766475d33

I don't see any link to it in current code. Could you link?

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content_data.ts:
Note that all of these are missing translations

nickmisasi replied:
I added translation strings but can't for the life of me remember what is the replacement for `t` - so `i18n-extract` isn't picking them up. Once I get that in place I think I'm good.

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.tsx:
```diff
@@ -0,0 +1,137 @@
+            onHide={onClose}
+            compassDesign={true}
+            enforceFocus={true}
+            ariaLabel='Cloud Preview Introduction'
```
This needs to be translatable either by translating it here or by referring to another element that contains this text with `aria-labelledby`

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.tsx:
```diff
@@ -0,0 +1,137 @@
+            footerContent={footerContent}
+            className='preview-modal-controller'
+        >
+            <button
```
Couldn't this use the built-in close button in the modal?

hmhealey replied:
That sounds like that might be an issue with the design then since the modal content has a title, so why aren't we putting that into the modal header? Looking at the screenshot above, I see that the SKU badge is above the title, so perhaps it could go below it?

This feels like a case where the design isn't following the standard pattern for modals which is leading to us having to jump through hoops with the code as well

CC @matthewbirtch

hmhealey reviewed (commented):
I started reviewing this before realizing that I was notified for some other reason, so feel free to disregard these

nickmisasi commented:
@hmhealey happy to have your review and would appreciate a full one if you've got the time. I had added codeowners but it didn't select a single person so I removed it. 

Re: refreshing bringing the modal back, this whole feature is gated around a new special license that only I have the ability to generate right now. Check is [here](https://github.com/mattermost/mattermost/pull/31235/files#diff-037a04515c4624a292194912dee2014f1e7a43a258e5873f3da76403eb0a3d29R24)

nickmisasi commented:
/update-branch

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.scss:
```diff
@@ -0,0 +1,151 @@
+        color: rgb(var(--center-channel-color-rgb));
+        font-size: 12px;
+        font-weight: 600;
+        gap: 8px;
```
```suggestion
        gap: 4px;
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.scss:
```diff
@@ -0,0 +1,151 @@
+        }
+    }
+
+    &__video-container {
+        display: flex;
+        overflow: hidden;
+        height: 306px;
+        flex: 0 0 306px;
+        align-items: center;
+        justify-content: center;
+        border-radius: 4px;
+        background: rgba(var(--center-channel-color-rgb), 0.04);
```
```suggestion
    &__media-container {
        display: flex;
        overflow: hidden;
        height: 306px;
        flex: 0 0 306px;
        align-items: center;
        justify-content: center;
        border: solid
        border-radius: 4px;
        border: solid 1px rgba(var(--center-channel-color-rgb), 0.08);
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.scss:
```diff
@@ -0,0 +1,151 @@
+
+    &__title {
+        min-height: 28px;
+        margin: 0 0 16px 0;
```
```suggestion
        margin: 0 0 8px 0;
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.scss:
```diff
@@ -0,0 +1,151 @@
+        display: flex;
+        width: fit-content;
+        align-items: center;
+        padding: 2px var(--spacing-xxxs, 4px);
```
I don't think we have these spacing variables in the codebase yet. Plus it should be 6px left/right padding

```suggestion
        padding: 2px 6px;
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.tsx:
```diff
@@ -0,0 +1,134 @@
+            <div className='preview-modal-controller__navigation-buttons'>
+                {isFirstSlide && (
+                    <button
+                        className='preview-modal-controller__skip-button'
```
```suggestion
                        className='btn btn-quaternary'
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.scss:
```diff
@@ -0,0 +1,152 @@
+        font-weight: 400;
+    }
+
+    &__navigation-buttons {
+        display: flex;
+        gap: 8px;
+    }
```
```suggestion
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.tsx:
```diff
@@ -0,0 +1,134 @@
+                        className='preview-modal-controller__nav-button preview-modal-controller__nav-button--tertiary'
+                        onClick={handlePrevious}
+                    >
+                        <ChevronLeftIcon size={18}/>
```
```suggestion
                        <ArrowLeftIcon size={18}/>
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.scss:
```diff
@@ -0,0 +1,152 @@
+        padding: 8px 16px;
+        border: 1px solid rgba(var(--center-channel-color-rgb), 0.16);
+        border-radius: 4px;
+        background: none;
+        color: var(--center-channel-color);
+        cursor: pointer;
+        font-size: 14px;
+        font-weight: 600;
+        gap: 4px;
+        transition: all 0.15s ease;
+
+        &:hover:not(:disabled) {
+            background: rgba(var(--center-channel-color-rgb), 0.08);
+        }
+
+        &:disabled {
+            cursor: not-allowed;
+            opacity: 0.5;
+        }
+
+        &--tertiary {
+            border: none;
+            background: rgba(var(--button-bg-rgb), 0.08);
+            color: var(--button-bg);
+
+            &:hover:not(:disabled) {
+                background: rgba(var(--button-bg-rgb), 0.16);
+            }
+        }
+
+        &--primary {
+            border: none;
+            background: var(--button-bg);
+            color: var(--button-color);
+
+            &:hover:not(:disabled) {
+                background: linear-gradient(0deg, rgba(0, 0, 0, 0.08) 0%, rgba(0, 0, 0, 0.08) 100%), var(--button-bg, #1C58D9);;
+            }
+        }
+    }
```
```suggestion
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.scss:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    .modal-content {
+        border-radius: 8px !important;
```
Should be able to remove this. Styles should be coming from modal.scss
```suggestion
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.scss:
```diff
@@ -0,0 +1,151 @@
+        border-radius: 4px;
+        margin-bottom: 12px;
+        background: rgba(var(--center-channel-color-rgb), 0.08);
+        color: rgb(var(--center-channel-color-rgb));
```
this should already be inherited, so don't likely need the `color`

matthewbirtch replied:
This color should already be inherited
```suggestion
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.scss:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    &__title {
+        min-height: 28px;
```
I don't think this is needed
```suggestion
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.scss:
```diff
@@ -0,0 +1,151 @@
+        min-height: 60px;
+        max-height: 80px;
+        margin-bottom: 24px;
+        color: rgb(var(--center-channel-color-rgb));
+        font-size: 14px;
+        line-height: 20px;
```
this should all be inherited
```suggestion
```

matthewbirtch commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_content.scss:
```diff
@@ -0,0 +1,151 @@
+            object-fit: contain;
+        }
+
+        iframe {
+            width: 100%;
+            height: 100%;
+            min-height: 300px;
+            align-self: stretch;
+            border: none;
+            border-radius: 8px;
+            box-shadow: 0 2px 8px rgba(var(--center-channel-color-rgb), 0.08);
+            transition: all 0.2s ease;
+
+            &:hover {
+                box-shadow: 0 4px 16px rgba(var(--center-channel-color-rgb), 0.12);
+                transform: translateY(-1px);
+            }
```
what is the iframe? Is this if youtube was used?

nickmisasi commented:
@matthewbirtch

1. I've adjusted the height - the 240px was from the base styling of GenericModal
2. Since it's loading the images from S3 on the fly, this is expected. It's the trade off between being able to ad-hoc change images as we wish, and having them be part of the bundle (requiring a PR/release). For now given the time constraints I would propose we leave it and potentially address later. 
3. I have commented out the slides for content flagging. For AI search, we don't actually mention it specifically in the title/subtitle. For now, we could simply change the image in the S3 bucket, as that image is what mentions agent search. This is the aformentioned advantage of loading on the fly
4. I totally missed that button. Ticket here: https://mattermost.atlassian.net/browse/CLD-9318 will have to address in a future PR.

matthewbirtch commented:
> * Since it's loading the images from S3 on the fly, this is expected. It's the trade off between being able to ad-hoc change images as we wish, and having them be part of the bundle (requiring a PR/release). For now given the time constraints I would propose we leave it and potentially address later.

Thanks @nickmisasi. Is there a way to cache the images. It reloads every time I'm cycling through the slides.

nickmisasi commented:
@matthewbirtch I've opened a ticket for this, will see what I can do.

hmhealey commented on webapp/channels/src/components/logged_in_route/index.tsx:
```diff
@@ -30,6 +31,7 @@ export default function LoggedInRoute(props: Props) {
                     {theme && (
                         <CompassThemeProvider theme={theme}>
                             <OnBoardingTaskList/>
+                            <CloudPreviewModal/>
```
Could you break this up so that there's a separate component that one component that just checks the license/subscription and then lazily loads the rest of the modal? As-is, even though we're using `makeAsyncComponent` above, we're still loading all of the data and CSS for this modal regardless of what the user's license is.

You're already sort of doing that with the `CloudPreviewModal` and `PreviewModalController` components, so I think the change would be to move the `makeAsyncComponent` into the `CloudPreviewModal` where it imports the `PreviewModalController` and refactor the modal slightly to move importing the data into the controller component

nickmisasi replied:
I've made some changes to address this. Lmk what you think.

hmhealey replied:
Than looks good to me. Thanks!

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.test.tsx:
```diff
@@ -0,0 +1,248 @@
+import PreviewModalController from './preview_modal_controller';
+
+// Mock the GenericModal component
+jest.mock('@mattermost/components', () => ({
```
Why mock the modal and icon components? Those are dumb enough that I wouldn't have expected us to need to mock those. The same goes for the MM logo component in the other test

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/cloud_preview_modal.tsx:
```diff
@@ -0,0 +1,72 @@
+
+const CLOUD_PREVIEW_MODAL_SHOWN_PREF = 'cloud_preview_modal_shown';
+
+const CloudPreviewModal: React.FC = () => {
```
Nit: I'm realizing that the names for this and the controller are sort of inverted because this outer component controls whether or not the modal is shown

nickmisasi replied:
I see what you're saying, but in this context by `controller` I meant it controls the functionality of the modal itself (ie, the slide show, the buttons within it etc). I'm open to other names but this its probably not a big deal to leave it?

hmhealey replied:
Typically, the various `WhateverController` components in the web app are either high level components in charge of loading and routing through parts of the app (`TeamController`, `ChannelController`, `LoginController`) or for determining when other elements should be mounted (`ModalController`, `AnnouncementBarController`). With the current naming, when I see that the `LoggedIn` component is rendering a `WhateverModal`, the first thing I think is "Why are we mounting a modal there? It shouldn't be mounted all the time."

Could you change those names around?

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.scss:
```diff
@@ -0,0 +1,93 @@
+.preview-modal-controller {
+    &.modal-dialog {
```
Are you sure we need all this extra, very-specific CSS for this modal? There's a lot of hard-coded dimensions that aren't responsive, and we're not making good use of what little shared components we have if we're manually needing to style so much of this since it's going to make maintenance a nightmare.

For shared elements like the close button specifically which I know I commented on before, perhaps we need some custom positioning, but we shouldn't need to respecify so much of how the button appears like the colours and cursor and such

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.test.tsx:
```diff
@@ -0,0 +1,248 @@
+
+    const renderComponent = (props = {}) => {
+        return render(
+            <IntlProvider locale='en'>
```
Could this use `renderWithContext` instead of just `render` with a manual `IntlProvider` in it?

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.test.tsx:
```diff
@@ -0,0 +1,248 @@
+    ];
+
+    beforeEach(() => {
+        mockOnClose.mockClear();
```
This isn't needed because we have Jest configured to automatically clear the mocks

hmhealey commented on webapp/channels/src/components/cloud_preview_modal/preview_modal_controller.test.tsx:
```diff
@@ -0,0 +1,248 @@
+                id: 'test.slide1.sku_label',
+                defaultMessage: 'First sku label',
+            },
+            videoUrl: 'https://www.youtube.com/watch?v=E3EGLxgNxNA',
```
Would it be possible to use a dummy URL for this? I don't know if Testing Library actually would render iframes, but it seems like it would be best to use something locally hosted to avoid having the tests be fragile

nickmisasi replied:
Since we're not actually using any YouTube videos right now and we can't actually dynamically set these until a future code change, I've opted to just remove the YouTube video support for now. If we do end up needing it in the future, I can add it back at that point.

gabrieljackson reviewed (approved):
LGTM
